common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java

```diff
@@ -290,8 +297,32 @@ public void childEvent(CuratorFramework curatorFramework, PathChildrenCacheEvent
         log.debug("Processing [{}] event for [{}]", pathChildrenCacheEvent.getType(), instance.getServiceId());
         switch (pathChildrenCacheEvent.getType()) {
             case CHILD_ADDED:
+                ScheduledFuture<?> task = delayedTasks.remove(instance.getServiceId());
+                if (task != null) {
+                    if (task.cancel(false)) {
+                        log.debug("[{}] Recalculate partitions ignored. Service was restarted in time [{}].",
+                                instance.getServiceId(), instance.getServiceTypesList());
+                    } else {
+                        log.debug("[{}] Going to recalculate partitions. Service was not restarted in time [{}]!",
+                                instance.getServiceId(), instance.getServiceTypesList());
+                        recalculatePartitions();
+                    }
+                } else {
+                    log.debug("[{}] Going to recalculate partitions due to adding new node [{}].",
+                            instance.getServiceId(), instance.getServiceTypesList());
+                    recalculatePartitions();
+                }
+                break;
             case CHILD_REMOVED:
-                recalculatePartitions();
+                ScheduledFuture<?> future = zkExecutorService.schedule(() -> {
+                    log.debug("[{}] Going to recalculate partitions due to removed node [{}]",
+                            instance.getServiceId(), instance.getServiceTypesList());
```

Comment:
Please, hold primitives in runnable, not instances. If it is small and immutable - it's ok as is